### PR TITLE
CCC21 hackathon: Manage kvm hosts with private key

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR provides a new option to add kvm host. 

It have the following changes:

(1) Add kvm host with private key (both API and UI changes) (https://github.com/apache/cloudstack/pull/5681/commits/ad59172c83cabd50f1cfa9fa5907be1025ffb864)

(2) cancel host maitenance by ssh to host using private key (https://github.com/apache/cloudstack/pull/5681/commits/135ed87a6557083eaab9aaea9d22663f8cdce319)

(3) update host with new private key (API changes) (https://github.com/apache/cloudstack/pull/5681/commits/2c566161ee2cd441e6aa24e77c81d59c1d2838dc)

(4) UI: add new dialog to update host password (https://github.com/apache/cloudstack/pull/5681/commits/d099b006a94d70e5431f5304415cd46ad775e46b)

(5) bug fix: cannot update host password on ubuntu 20.04 (https://github.com/apache/cloudstack/pull/5681/commits/1e2a09da54f735601cded2875cc0f4a28edd2e52)

(6) bug fix: new password is always applied on host when update host password (https://github.com/apache/cloudstack/pull/5681/commits/3803969239401038a65bba91ec0569d7b74ef822)

Thanks @rhtyd for your good idea !

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### Screenshots (if appropriate):

1. add host

image

2. update host password
image

### How Has This Been Tested?

1. added host, worked as expected

(1) private key
(2) password
(3) private key and (correct, wrong) password

2. put host to maitenance, stop cloudstack-agent

(1) cancel maintenance worked as expected.
(2) update host with wrong password, cancel maintenance worked as expected.
(3) update host with correct password and wrong private key, cancel maintenance worked as expected.
(4) update host with wrong password and wrong private key, cancel maintenance failed as expected.

3. removed host, worked as expected

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/command/admin/host/AddHostCmd.java [hunk 1]
```diff
@@ -60,6 +60,9 @@
     @Parameter(name = ApiConstants.PASSWORD, type = CommandType.STRING, description = "the password for the host; required to be passed for hypervisors other than VMWare")
     private String password;
 
+    @Parameter(name = ApiConstants.PRIVATE_KEY, type = CommandType.STRING, description = "the private key for the host; support KVM only.", length = 16384)
+    private String privateKey;
```
Comment (rhtyd) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/host/AddHostCmd.java:64]: No this beats the purpose, we should assume that admin will setup and sync mgmt server ssh key for the cloud user across hosts.

**Comment by rhtyd** [comment]
Thanks @weizhouapache for submitting, left one remark. The idea was primarily that people don't want to share credentials with CloudStack (either passwords or pub/priv key)

**Comment by weizhouapache** [comment]
> Thanks @weizhouapache for submitting, left one remark. The idea was primarily that people don't want to share credentials with CloudStack (either passwords or pub/priv key)

@rhtyd 
we need to figure out which username and private key file will be used. 
root user and /var/lib/cloudstack/management/.ssh/id_rsa ?

**Comment by weizhouapache** [comment]
this is not same as what Rohit purposed.
closing this pr then.

**Comment by rhtyd** [comment]
@weizhouapache Yes I think it should be management server's cloud user ssh key which is stored in the database too. We may show like a UI form to tell admin the public ssh key which they can copy to authorized_keys file of kvm host for the root user. This is generally the practice I think, we can ask community for feedback @wido @GabrielBrascher @DaanHoogland @mlsorensen @kiwiflyer @nvazquez @pdion891 @svenvogel et al.

Alternatively, we can display steps in the UI for the admin to run specific shell or run cloudStack-setup-agent commands to allow for out of band adding of hosts (manually for admin who are very security-risk averse)

**Comment by weizhouapache** [comment]
> @weizhouapache Yes I think it should be management server's cloud user ssh key which is stored in the database too. We may show like a UI form to tell admin the public ssh key which they can copy to authorized_keys file of kvm host for the root user. This is generally the practice I think, we can ask community for feedback @wido @GabrielBrascher @DaanHoogland @mlsorensen @kiwiflyer @nvazquez @pdion891 @svenvogel et al.
> 
> Alternatively, we can display steps in the UI for the admin to run specific shell or run cloudStack-setup-agent commands to allow for out of band adding of hosts (manually for admin who are very security-risk averse)

@rhtyd 
let's leave the option to users. 
They can decide whether they want to sync /var/lib/cloudstack/management/.ssh/id_rsa.pub to kvm hosts (it is also saved in cloudstack DB).

I will create another PR, it will be easier.

**Pull request closed without merging**